dlrover/python/elastic_agent/torch/ckpt_saver.py

```diff
@@ -722,91 +797,139 @@ def _commit_checkpoint(
-
-        if not all(acquired):
-            for i, lock in enumerate(self._shm_lock):
-                if acquired[i]:
-                    lock.release()
-            return False
 
-        return True
+class CheckpointEngine(metaclass=ABCMeta):
+    def __init__(self, checkpoint_dir):
+        self.checkpoint_dir = checkpoint_dir
+        if dist.is_initialized():
+            self._rank = dist.get_rank()
+            self._local_rank = int(os.environ["LOCAL_RANK"])
+        else:
+            self._rank = 0
+            self._local_rank = int(os.getenv("LOCAL_RANK", 0))
+        self._saver_group = None
+        self._buffer_size = 0
+        self._cached_step = 0
+        self._restart_count = env_utils.get_torch_restart_count()
+        # queue for agent to save to storage, only rank 0
+        if self._rank == 0:
+            self._to_save_queue = SharedQueue(
+                name=_SAVE_STEP_QNAME_PREFIX + str(0), create=False
+            )
+        else:
+            self._to_save_queue = None
+        # lock for shared memory
+        local_shard_num = self.get_local_shard_num()
+        self.local_shard_id = self._local_rank % local_shard_num
+        lock_name = _SHM_LOCK_NAME_PREFIX + str(self.local_shard_id)
+        self._shm_lock = SharedLock(name=lock_name, create=False)
+        self._shm_handler = SharedMemoryHandler(
+            self.local_shard_id, host=False
+        )
+        self._notify_agent_to_create_saver()
 
-    def save_shm_to_storage(self):
-        """
-        Save the state dict in the shared memory into the storage. The agent
-        can call the method to save the state dict into the storage if the
-        training process fails or the agent wants to restart training
-        processes.
-        """
-        logger.info("Agent terminated, save latest checkpoint to storage")
-        for handler in self._shm_handlers:
-            if handler.empty():
-                return
+    def __del__(self):
+        self.close()
 
-        if self._writing_storage:
-            logger.info("Saver is writing to storage, waiting...")
-            start = time.time()
-            while self._writing_storage:
-                time.sleep(10)
-                elapsed_time = time.time() - start
-                if elapsed_time > 120:
-                    logger.error("Saver writing to storage, timeout")
-                    return
+    def close(self):
+        self._shm_handler.close()
 
-        if not self._acquire_all_locks():
-            # The training process does not release the lock because it fails
-            # when writing the state dict into the shared memory. The shared
-            # memory may be dirty and the saver cannot save it to the storage.
+    def _notify_agent_to_create_saver(self):
+        if self._local_rank != 0:
             return
+        if self._restart_count > 0:
+            # Only local rank 0 notify to initialize the saver in
```

Comment:
since the saver only initialized once, if an agent start with a new global_shard_num in elastic training, we can't reset this parameter, this may cause problem when saving to storage.

Reply:
we may change the lifecycle of Saver，one instance for each training attempt。

Reply:
It is difficult to re-initialize a new Saver after the Saver starts the thread to wait the event to save.

Reply:
We cannot re-create a new Saver because the new Saver will clean the tensor meta in the shared dict and the shared memory  with or without the change of `global_shard_num`. If the `global_shard_num` is not changed, the training need to restore the checkpointing state dict from the shared memory with tensor meta.